### PR TITLE
Detect and warn when source files are out of date

### PR DESCRIPTION
**Pull request by andrewcrawley**
When we hit a breakpoint, warn the user if the source file is newer than the module containing it.

Please review: @jacdavis @gregg-miskelly @chuckries @paulmaybee

### CODE REVIEW

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 1]
```diff
@@ -744,8 +753,17 @@ private void Dispose()
...
                 ulong addr = cxt.pc ?? 0;
 
-                bool fContinue;
                 TupleValue frame = results.Results.TryFind<TupleValue>("frame");
+                if (frame != null && addr != 0)
+                {
+                    string sourceFile = frame.TryFindString("fullname");
+                    if (!String.IsNullOrEmpty(sourceFile))
+                    {
+                        await this.VerifySourceFileTimestamp(addr, sourceFile);
```
Comment (paulmaybee) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: I don't think you want to do this check unless the debugger is really going to stop. Maybe delay this check until the decision to send up a stopping event is made.

**Comment by paulmaybee** [comment]
👍

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs (hunk 1 above)
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: This is also only going to do the verification in the leaf stack frame. We might want to check other frames on the call stack.

---
In reply to: [62338268](https://github.com/Microsoft/MIEngine/pull/308#discussion_r62338268) [](ancestors = 62338268)

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 2]
```diff
@@ -858,6 +876,40 @@ private void Dispose()
...
+            await this.EnsureModulesLoaded();
+
+            string targetModulePath = this._launchOptions.ExePath;
+            DebuggedModule targetModule = this._moduleList.FirstOrDefault(m => m.AddressInModule(addr));
+            if (targetModule != null)
+            {
+                targetModulePath = targetModule.Name;
+            }
+
+            if (!File.Exists(sourceFilePath) || !File.Exists(targetModulePath))
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Don't we need also need --
    targetModulePath == null ||

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 3]
```diff
@@ -858,6 +876,40 @@ private void Dispose()
...
+            DateTime sourceFileTimestamp = File.GetLastWriteTimeUtc(sourceFilePath);
+            DateTime moduleFileTimestamp = File.GetLastWriteTimeUtc(targetModulePath);
+
+            if (sourceFileTimestamp > moduleFileTimestamp)
+            {
+                // Source file is newer than the module - warn the user if we haven't already
+				Tuple<string, string> key = Tuple.Create(sourceFilePath, targetModulePath);
+
+                if (!this._fileTimestampWarnings.Contains(key))
+                {
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Should we check this before asking the file system for the time?

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 4]
```diff
@@ -858,6 +876,40 @@ private void Dispose()
             }
         }
 
+        private async Task VerifySourceFileTimestamp(ulong addr, string sourceFilePath)
+        {
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: No nothing if we are using pipe transport?

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 5]
```diff
@@ -858,6 +876,40 @@ private void Dispose()
...
+                targetModulePath = targetModule.Name;
+            }
+
+            if (!File.Exists(sourceFilePath) || !File.Exists(targetModulePath))
+            {
+                return;
+            }
+
+            DateTime sourceFileTimestamp = File.GetLastWriteTimeUtc(sourceFilePath);
+            DateTime moduleFileTimestamp = File.GetLastWriteTimeUtc(targetModulePath);
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Do we want a try/catch around these to guard against I/O errors?

**Comment by gregg-miskelly** [comment]
Otherwise looks good to me.

**Comment by jacdavis** [comment]
Gregg already covered everything I would have said on this.

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs (hunk 2 above)
Comment (andrewcrawley) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: File.Exists will return false if you pass null / empty string, so this is handled.

Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Cool. You learn something new everyday :)

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 6]
```diff
@@ -858,6 +876,49 @@ private void Dispose()
...
+            {
+                if (!File.Exists(sourceFilePath) || !File.Exists(targetModulePath))
+                {
+                    return;
+                }
+
+                DateTime sourceFileTimestamp = File.GetLastWriteTimeUtc(sourceFilePath);
+                DateTime moduleFileTimestamp = File.GetLastWriteTimeUtc(targetModulePath);
+
+                if (sourceFileTimestamp > moduleFileTimestamp)
```
Comment (faxue-msft) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Do we need to also warn user if the source file is older?

Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: No. That would be the normal case.

**Pull request merged**